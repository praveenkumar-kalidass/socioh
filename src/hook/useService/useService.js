import * as Keychain from 'react-native-keychain';
import AsyncStorage from '@react-native-community/async-storage';
import RNContacts from 'react-native-contacts';

import { CONSTANT, ERROR, TRANSLATION } from '../../constant';
import useAjax from '../useAjax';
import useUser from '../useUser';

const useService = () => {
  const { ajax } = useAjax();
  const { user, saveUserDetails, clearUserDetails } = useUser();

  const signUp = async ({ name, email, password }) => {
    await Keychain.resetGenericPassword();
    await Keychain.setGenericPassword(email, password);
    await AsyncStorage.setItem(
      CONSTANT.ASYNC_STORAGE_KEY.SOCIOH_USER,
      JSON.stringify({
        name,
        email,
      }),
    );
    saveUserDetails({
      name,
      email,
    });
    return ajax();
  };

  const signIn = async ({ email, password }) => {
    const credentials = await Keychain.getGenericPassword();
    if (credentials.username === email && credentials.password === password) {
      const userDetails = await AsyncStorage.getItem(
        CONSTANT.ASYNC_STORAGE_KEY.SOCIOH_USER,
      );
      saveUserDetails(JSON.parse(userDetails));
      return ajax();
    }
    throw Error(ERROR.INVALID_CREDENTIALS);
  };

  const logout = async () => {
    clearUserDetails();
    return ajax();
  };

  const getUserDetails = async () => {
    if (user) {
      await ajax();
      return {
        ...user,
        interests: Object.values(TRANSLATION.INTERESTS),
      };
    }
    throw Error(ERROR.NO_USER_FOUND);
  };

  const getContacts = async (search) => {
    const permission = await RNContacts.requestPermission();
    if (permission === CONSTANT.PERMISSION.AUTHORIZED) {
      let contacts;
      if (search) {
        contacts = await RNContacts.getContactsMatchingString(search);
      } else {
        contacts = await RNContacts.getAll();
      }
      await ajax();
      return contacts;
    }
    throw Error(ERROR.UNAUTHORIZED_TO_READ_CONTACTS);
  };

  const getMessages = async () => {
    await ajax();
    return CONSTANT.MESSAGES.GREETINGS;
  };

  const getMessageList = async () => {
    await ajax();
    return CONSTANT.MESSAGES.LIST;
  };

  return {
    signUp,
    signIn,
    logout,
    getUserDetails,
    getContacts,
    getMessages,
    getMessageList,
  };
};

export default useService;
