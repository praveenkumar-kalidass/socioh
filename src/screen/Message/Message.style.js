import styled from 'styled-components/native';
import { Button } from 'native-base';

import { COLOR, SPACING } from 'constant';

export const MessageContainer = styled.View`
  flex: 1;
`;

export const MessageGrid = styled.ScrollView`
  flex: 1;
  padding-horizontal: ${SPACING.SPACE_20};
`;

export const InputContainer = styled.View`
  background-color: ${COLOR.WHITE};
`;

export const SendButton = styled(Button)`
  background-color: ${COLOR.PRIMARY};
  margin-vertical: ${SPACING.SPACE_5};
`;
