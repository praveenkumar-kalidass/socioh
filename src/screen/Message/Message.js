import React, { useEffect, useState } from 'react';
import { Input, Item, Icon } from 'native-base';

import { ICON, TRANSLATION } from '../../constant';
import MessageTile from '../../component/MessageTile';
import useService from '../../hook/useService';
import {
  InputContainer,
  MessageContainer,
  MessageGrid,
  SendButton,
} from './Message.style';

const Message = () => {
  const [chat, setChat] = useState([]);

  const { getMessages } = useService();

  useEffect(() => {
    const loadMessages = async () => {
      try {
        const messages = await getMessages();
        setChat(messages.map((message) => ({ text: message, received: true })));
      } catch (error) {
        console.warn('Error:', error);
      }
    };

    loadMessages();
  }, []);

  return (
    <MessageContainer>
      <MessageGrid>
        <For each="message" index="index" of={chat}>
          <MessageTile
            key={`message_${index}`}
            testID={`message_${index}`}
            message={message.text}
            isReceived={message.received}
          />
        </For>
      </MessageGrid>
      <InputContainer>
        <Item regular>
          <Input placeholder={TRANSLATION.WRITE_YOUR_MESSAGE} />
          <SendButton>
            <Icon type={ICON.FONT_AWESOME_FAMILY} name={ICON.SEND} />
          </SendButton>
        </Item>
      </InputContainer>
    </MessageContainer>
  );
};

export default Message;
