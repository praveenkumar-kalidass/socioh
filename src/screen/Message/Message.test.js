import React from 'react';
import { act, fireEvent, render, waitFor } from '@testing-library/react-native';

import useService from '../../hook/useService';
import Message from './index';

jest.mock('../../hook/useService');

describe('Message', () => {
  const mockMessagesService = jest.fn();

  beforeEach(() => {
    mockMessagesService.mockResolvedValue(['Hi']);
    useService.mockImplementation(() => ({ getMessages: mockMessagesService }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const container = render(<Message />);

    expect(container).toMatchSnapshot();
  });

  it('should load messages when screen is loaded', async () => {
    const { getByTestId } = render(<Message />);

    await waitFor(() => getByTestId('message_0'));

    expect(getByTestId('message_0').props.children).toStrictEqual('Hi');
  });

  it('should send message when send button is clicked with message', async () => {
    const { getByTestId } = render(<Message />);

    await waitFor(() => getByTestId('message_0'));

    await act(async () => {
      await fireEvent.changeText(getByTestId('message_input'), 'Hello');
      await fireEvent.press(getByTestId('message_send'));
    });

    expect(getByTestId('message_0').props.children).toStrictEqual('Hi');
    expect(getByTestId('message_1').props.children).toStrictEqual('Hello');
    expect(getByTestId('message_2').props.children).toStrictEqual('Pardon?');
  });
});
