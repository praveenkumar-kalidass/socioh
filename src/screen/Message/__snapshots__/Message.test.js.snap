// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "paddingHorizontal": 20,
        },
      ]
    }
  >
    <View
      maxWidth={375}
      received={true}
      style={
        Array [
          Object {
            "alignSelf": "flex-start",
            "backgroundColor": "#1abfbf",
            "borderBottomRightRadius": 5,
            "borderTopLeftRadius": 5,
            "borderTopRightRadius": 5,
            "marginBottom": 10,
            "maxWidth": 375,
            "paddingBottom": 10,
            "paddingLeft": 10,
            "paddingRight": 10,
            "paddingTop": 10,
          },
        ]
      }
    >
      <View
        received={true}
        style={
          Array [
            Object {
              "bottom": 0,
              "left": -10,
              "position": "absolute",
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          received={true}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 12,
              },
              Array [
                Object {
                  "color": "#000",
                  "fontSize": 30,
                },
                Object {
                  "color": "#000",
                  "fontSize": 30,
                },
                Object {
                  "color": "#1abfbf",
                },
              ],
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
          type="FontAwesome"
        >
          
        </Text>
      </View>
      <Text
        received={true}
        style={
          Array [
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            },
            Object {
              "color": "#ffffff",
            },
          ]
        }
        testID="message"
      >
        Hi, How are you?
      </Text>
    </View>
    <View
      maxWidth={375}
      received={false}
      style={
        Array [
          Object {
            "alignSelf": "flex-end",
            "backgroundColor": "#feda68",
            "borderBottomLeftRadius": 5,
            "borderTopLeftRadius": 5,
            "borderTopRightRadius": 5,
            "marginBottom": 10,
            "maxWidth": 375,
            "paddingBottom": 10,
            "paddingLeft": 10,
            "paddingRight": 10,
            "paddingTop": 10,
          },
        ]
      }
    >
      <View
        received={false}
        style={
          Array [
            Object {
              "bottom": 0,
              "position": "absolute",
              "right": -10,
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          received={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 12,
              },
              Array [
                Object {
                  "color": "#000",
                  "fontSize": 30,
                },
                Object {
                  "color": "#000",
                  "fontSize": 30,
                },
                Object {
                  "color": "#feda68",
                },
              ],
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
          type="FontAwesome"
        >
          
        </Text>
      </View>
      <Text
        received={false}
        style={
          Array [
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            },
            Object {
              "color": "#000000",
            },
          ]
        }
        testID="message"
      >
        Hi, How are you?
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ffffff",
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={false}
      forwardedRef={[Function]}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderBottomWidth": 1,
          "borderColor": "#D9D5DC",
          "borderLeftWidth": 1,
          "borderRadius": undefined,
          "borderRightWidth": 1,
          "borderTopWidth": 1,
          "borderWidth": undefined,
          "flexDirection": "row",
          "marginLeft": 2,
          "opacity": 1,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        editable={true}
        placeholder="Write your message here..."
        placeholderTextColor="#575757"
        rejectResponderTermination={true}
        style={
          Object {
            "color": "#000",
            "flex": 1,
            "fontSize": 17,
            "height": 50,
            "paddingLeft": 8,
            "paddingRight": 5,
            "top": 1.5,
          }
        }
        underlineColorAndroid="rgba(0,0,0,0)"
      />
      <View
        accessible={true}
        focusable={false}
        forwardedRef={[Function]}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "flex-start",
            "backgroundColor": "#1abfbf",
            "borderBottomWidth": null,
            "borderColor": "#007aff",
            "borderLeftWidth": null,
            "borderRadius": 5,
            "borderRightWidth": null,
            "borderTopWidth": null,
            "borderWidth": undefined,
            "elevation": 2,
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "space-between",
            "marginVertical": 5,
            "opacity": 1,
            "paddingBottom": 6,
            "paddingTop": 6,
            "shadowColor": undefined,
            "shadowOffset": undefined,
            "shadowOpacity": undefined,
            "shadowRadius": undefined,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 12,
              },
              Array [
                Object {
                  "color": "#000",
                  "fontSize": 30,
                },
                Object {
                  "color": "#fff",
                  "fontSize": 24,
                  "marginLeft": 16,
                  "marginRight": 16,
                  "paddingTop": 2,
                },
              ],
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
          type="FontAwesome"
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;
